airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py

```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+import time
+from concurrent.futures import Future
+from typing import Any, List
+
+
+class Throttler:
+    """
+    A throttler that waits until the number of concurrent tasks is below a certain threshold.
+    """
+
+    def __init__(self, futures_list: List[Future[Any]], sleep_time: float, max_concurrent_tasks: int):
+        """
+        :param futures_list: The list of futures to monitor
```

Comment:
Maybe this is obvious, but why do we need to throttle workers enqueuing items based on the number of pending tasks? I'm wondering if we can solve some of these memory issues by setting [`queue.maxsize`](https://docs.python.org/3/library/queue.html#queue.Queue), which prevents items from being added until something else is dequeued, creating a sort of back pressure.

Reply:
Good question!

The issue I ran into with blocking on the size of the queue is that the main thread will remove elements from the queue before potentially adding them to the list of futures. Since the main thread doesn't wait, it'll be able to remove items from the queue even if the list of futures is large, so the tasks won't wait.

![Screenshot 2024-01-16 at 2 16 22 PM](https://github.com/airbytehq/airbyte/assets/1451943/b1d94b2d-3d65-460a-89ab-e7a06ec36b5d)

Reply:
Cool, thanks for the clarification! I think we should document the reason that we need this. Mind adding a docstring on this file?